### PR TITLE
*: Bump async-std to v1.6.5

### PR DESCRIPTION
**Pull request by mxinden**
Prevent users from using v1.6.4 which faces issues receiving incoming
TCP connections. See https://github.com/async-rs/async-std/issues/888
for details.

//CC @JoshOrndorff

Release note suggestion:

> Methods on `sc_network::gossip::QueuedSender` take `&mut self` instead of `self`.

### CODE REVIEW

**Comment by bkchr** [comment]
As we had seen last week in polkadot, this makes the web wasm build fail.

**Comment by mxinden** [comment]
> As we had seen last week in polkadot, this makes the web wasm build fail.

This would be fixed with 96c9f6b which uses a channel instead of a condvar as suggested [here](https://github.com/paritytech/substrate/pull/6803#pullrequestreview-466063854):

```
client/network/src/gossip: Use channel instead of condvar

`async_std::sync::Condvar::wait_timeout` uses
`gloo_timers::callback::Timeout` when compiled for
`wasm32-unknown-unknown`. This timeout implementation does not fulfill
the requirement of being `Send`.

Instead of using a `Condvar` use a `futures::channel::mpsc` to signal
progress from the `QueuedSender` to the background `Future`.
```

See https://github.com/rustwasm/gloo/issues/109 for details on `Timeout` not being `Send`.

Before I iron out any outstanding `TODO`s, @tomaka would this be a valid solution for you?

File: client/network/src/gossip.rs [hunk 1]
```diff
@@ -67,6 +67,9 @@ mod tests;
 pub struct QueuedSender<M> {
 	/// Shared between the front and the back task.
 	shared: Arc<Shared<M>>,
+	// TODO: Can we get around this Mutex? E.g. by taking `&mut` for `lock_queue` and
+	// `queue_or_discard`.
+	notify_background_future: Mutex<Sender<()>>,
```
Comment (mxinden) [inline client/network/src/gossip.rs:72]: @tomaka I would like to remove the need for this `Mutex`, thus requiring `QueuedSender::lock_queue` and `QueuedSender::queue_or_discard` to take a `&mut self`. Is the current `&self` a design requirement?

**Review by mxinden — COMMENTED** [review]
> tomaka added the A1-needsburnin label 43 minutes ago

I don't  think the `needsburnin` label is helpful here. As far as I can tell `QueuedSender` is never constructed within the Substrate or Polkadot codebase apart from unit tests.

File: client/network/src/gossip.rs [hunk 2]
```diff
@@ -180,66 +173,52 @@ impl<'a, M: Send + 'static> QueueGuard<'a, M> {
...
-			Err(_) => continue,
-		};
-
-		let ready = match sender.ready().await {
-			Ok(r) => r,
-			Err(_) => continue,
-		};
-
-		let _ = ready.send(messages_encode(next_message));
+		if wait_for_sender.next().await.is_none() {
```
Comment (tomaka) [inline client/network/src/gossip.rs:197]: I don't think that works. You're blocking indefinitely here.

Comment (mxinden) [reply client/network/src/gossip.rs:197]: Can you add more details why that is a problem @tomaka? Reaching this line implies that there are currently no more messages to send. Thus blocking until the sender adds more messages should be the right approach.

Comment (tomaka) [reply client/network/src/gossip.rs:197]: I see, I totally misread the code above. Nevermind.

**Comment by mxinden** [comment]
@tomaka would you mind taking another look?

**Review by tomaka — APPROVED** [review]

**Comment by mxinden** [comment]
This patch has been deployed on two sentry nodes for the last 3 hours. _Substrate service tasks_, _Substrate networking_ nor the _Node Exporter_ dashboard show any anomalies.

Only thing sticking out is the number of peers the nodes have more than one connection to. This seems to be increasing since the redeployment of the nodes:

image

**Comment by mxinden** [comment]
bot merge

**Pull request merged**